scfinder: adding dynamic window length for envelope maxima

jenrandrews opened the pull request:
Added the option of using a dynamic time interval when selecting PGA maxima from the ring buffer to pass to FinDer. This allows use of a smaller time interval by default that is less likely to combine random noise spikes but allows for average station spacing. When there is an active FinDer event, the time window will be increased (up to a maximum of the total buffer length) based on the event rupture length/magnitude, which allows for the rupture duration.

Documentation for new configuration parameters added.

Testing done to verify window used to pass values to FinDer resizes and resets as expected, and total ring buffer length remains unchanged as expected. Docker build and run checked as per https://github.com/SED-EEW/SED-EEW-SeisComP-contributions/discussions/5.

FMassin commented:
Hi @jenrandrews,

As it has been tested, and if your are happy with the output, it should be merged. But can you fix the indentation? And can you attach here the logs of your test mentioning the window length changes?

jenrandrews commented:
hi @FMassin - thanks for looking at this so quickly! 

I've fixed the indentation as requested, this should now show in the pull request diffs.

Regarding the testing I don't have all of the log files, but I carried out the following checks using replays of Kaikoura M7.8:
- checked PGA ring buffer's "earliest valid timestamp" computed using envelopeBufferSize (_bufferLength), and made sure it didn't change as envelopeDefaultBufLen (_bufDefaultLen) was varied;
- set envelopeDefaultBufLen to 10s, 30s, 60s and 120s and verified that event results always reached max magnitude 7.5, and verified that internal variable _bufVarLen (used to get PGAs for FinDer at each timestep) was set to default and then modified as event rupture updated;
- forced internal variable _bufVarLen to stay fixed at 10s and verified that event max magnitude was under-estimated;
- forced event hold time to expire and verified that internal variable _bufVarLen was reset to _bufDefaultLen.

Example of new log lines during event replay:
>> grep "Increasing" ../../rcet_playback/Kaikoura_MS_7.8/dot_seiscomp3/log/scfinder.log
2016/11/13 11:03:54 [debug/SCFINDER] Increasing data window to 65 because of active FinDer event rupture length 98.5
2016/11/13 11:03:56 [debug/SCFINDER] Increasing data window to 65 because of active FinDer event rupture length 98.5
2016/11/13 11:03:57 [debug/SCFINDER] Increasing data window to 65 because of active FinDer event rupture length 98.5
2016/11/13 11:03:59 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:00 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:01 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:02 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:03 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:05 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:06 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:07 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:08 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:09 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:11 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:12 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:13 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:14 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:15 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:16 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:17 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:18 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:20 [debug/SCFINDER] Increasing data window to 76 because of active FinDer event rupture length 114.9
2016/11/13 11:04:22 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:23 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:24 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:26 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:27 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:33 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:34 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:36 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:37 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:38 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:40 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:41 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:43 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:45 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:46 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1
2016/11/13 11:04:47 [debug/SCFINDER] Increasing data window to 89 because of active FinDer event rupture length 134.1

Let me know if anything further is needed.
Cheers,
Jen

FMassin commented:
Awesome work @jenrandrews, thank you so much.
In practice to avoid false trigger and use this new feature when a true event occurs, I understand that a value smaller than envelopeBufferSize  should be used in envelopeDefaultBufLen ? If so could you suggest a generic envelopeDefaultBufLen value that you could set as default in xml and cpp files? Would 10s be too short?

jenrandrews commented:
Thanks for the suggestions @FMassin! 
I've updated the variable name, changed the default for "defaultFinDerEnvelopeLength" to 60s, and added a note that the value needs to be based on station spacing and number of FinDer trigger stations. 

For the default for defaultFinDerEnvelopeLength, I would be equally happy to return that to 120s and effectively switch the behaviour off using the default values, in order to mimic the existing behaviour with only envelopeBufferSize available. This would be sensible if there are any users who might inadvertently update and not realise the change in configuration - would there be any users like that?

As points of reference, in ShakeAlert envelopeBufferSize is currently 180s, but is about to be increased to 300s to account for M9s in Cascadia. With triggering stations set to 4, the defaultFinDerEnvelopeLength is set to 60s. The latter is very generous for southern California where the network is dense (and a value of 30s would be easily sufficient), but is conservative in case of events in sparse areas of OR and WA states. For NZ I think I will suggest envelopeBufferSize 180s and defaultFinDerEnvelopeLength of 60s or 90s for 3 triggering stations, as there are parts of the NZ network that are sparse.

FMassin commented:
OK! I'll merge when the build check completes with my little mention in the changelog. Thanks again!
